Slot mappings documentation update (#10011)
RasaHQ/rasa: docs/docs/domain.mdx

  - action: action_explain_table_limit
```

### Slot Mappings

:::info New in 3.0

As of 3.0, slot mappings are defined in the `slots` section of the domain.

> This change removes the implicit mechanism of setting slots via auto-fill which is replaced with a new explicit
> mechanism of setting slots after every user message.

Reviewer: ```suggestion
This change removes the implicit mechanism of setting slots via auto-fill and replaces it with a new explicit
mechanism of setting slots after every user message.
```

      value: my_value
      intent: intent_name
      not_intent: excluded_intent
      conditions:
      - active_loop: your_form
forms:
  your_form:
    required_slots:
    - slot_name
```

### Mapping Conditions

To apply a slot mapping only within the context of a form, specify
the name of the form in the `conditions` key of a slot mapping. Conditions list the form name(s)
for which the mapping is applicable in the `active_loop` key.

Conditions can also include the name of the `requested_slot`. If `requested_slot` is not mentioned,
then the slot will be set if relevant information is extracted, regardless of which slot is being
requested by the form.

```yaml-rasa
slots:
  slot_name:
    type: text
    mappings:
    - type: from_text
      intent: intent_name
      conditions:
      - active_loop: your_form
        requested_slot: slot_name
      - active_loop: another_form
```

:::note
If `conditions` are not included in a slot mapping, the slot mapping will be applicable regardless of whether
any form is active. As long as a slot is listed in a form's `required_slots`, the form will prompt for the slot
if it is empty when the form is activated.
:::

### Custom Slot Mappings

> 
> You can define custom slot mappings using [slot validation actions](./slot-validation-actions.mdx) when none of the
> predefined mappings fit your use case. You must define this slot mapping to be of type `custom`, for example:
> 
> ```yaml
> slots:
>   day_of_week:
>     type: text
>     mappings:
>     - type: custom
>       action: action_calculate_day_of_week

Reviewer 2: ```suggestion

You can define custom slot mappings using [slot validation actions](./slot-validation-actions.mdx) when none of the
predefined mappings fit your use case. You must define this slot mapping to be of type `custom`, for example:

```yaml
slots:
  day_of_week:
    type: text
    mappings:
    - type: custom
      action: action_calculate_day_of_week
\```

You can also use the `custom` slot mapping to list slots that will be filled by arbitrary custom actions in the course of a conversation, by listing the type and no specific. For example:

```yaml
slots:
  handoff_completed:
    type: boolean
    mappings:
    - type: custom
\```

This slot will not be updated on every user turn, but only once a custom action that returns a `SlotSet` event for it is predicted. 

```

Can't get the backticks to play along; Is this true though? To maintain the previous behaviour of a slot that is created without being filled on each turn user's will need to know this

Author: Yup, that's correct, I'll add that 👍🏼

### Slot Auto-fill

> :::caution Removed in 3.0

Reviewer: We don't have a clear guideline for this at the moment. The only guideline we have now is that we remove callouts when we remove deprecated features / behaviours in order to start fresh in a new major and reducing the amount of callouts in a doc page (i.e. it doesn't scale otherwise). I would assume the same for this case (even if nothing was deprecated before).

How about:
1. remove this section altogether as it doesn't apply to 3.x
2. in the callout that announces the new feature (i.e. ```:::info New in 3.0```), use some of the text that you have here to briefly explain the behaviour change and link at the migration guide? (like @melindaloubser1  suggested)

?

Reviewer: I would add such callout in [the next section below](https://github.com/RasaHQ/rasa/pull/10011/files#diff-cb86102a1484b02650a25eceae48c41c59f43802ab70061e6dd1d120327cd0e6R483-R484), as well as in `docs/docs/forms.mdx`. I think it's good to add duplicated callouts for the same feature wherever we need them (i.e. across different docs pages), because it's more explicit to the user!

🤔 I think I should add this as a documentation guideline, WDYT?

Author: Sounds good, thank you 👍🏼

Slot Auto-fill functionality was removed in 3.0 and replaced with a new explicit mechanism of
setting slots after every user message.
You will need to explicitly define slot mappings for each slot in the `slots` section of `domain.yml`.
In addition, stories must include the `slot_was_set` step.
Please read through the migration guide to update your assistant.

> 

Author: Actually L478 (on `stories must include the slot_was_set step`) won't be a requirement until a later minor, considering this long thread [here](https://rasa-hq.slack.com/archives/C01HHMR4X8S/p1635861549213800). I will remove it and add the link to migration guide as you suggested.

:::

### Slot Mappings

Rasa Open Source comes with four predefined mappings to fill the slots of a form

> based on the latest user message.

Author: Good catch 😄

      entity: entity_name
      role: role_name
      group: group name
      intent: intent_name
      not_intent: excluded_intent
```

:::note
Within the context of a form, a slot with `from_entity` mapping will be set only if the mapping
is unique, unless the slot is requested. You can find out more in [Forms](./forms.mdx#unique-entity-mapping-matching).
:::

#### from_text

The `from_text` mapping will use the text of the next user utterance to fill the slot
`slot_name`. If `intent_name` is `None`, the slot will be filled regardless of intent name.
Otherwise, the slot will only be filled if the user's intent is `intent_name`.

The slot mapping will not apply if the intent of the message is `excluded_intent`.
Note that you can define lists of intents for the parameters `intent` and `not_intent`.

```yaml-rasa
slots:
  slot_name:
    type: text
    mappings:
    - type: from_text
      intent: intent_name
      not_intent: excluded_intent
```

#### from_intent

The `from_intent` mapping will fill slot `slot_name` with value `my_value` if
user intent is `intent_name` or `None`. The slot mapping will not
apply if the intent of the message is `excluded_intent`. Note that you can

> also define lists of intents for the parameters `intent` and `not_intent`.

Author: Hmm, I am not the original author of this section, but based on the (translated) code in [here](https://github.com/RasaHQ/rasa/blob/1ab2908e9e8dca8cf140220f3cbd9c4fb0acbc6e/rasa/shared/core/slot_mappings.py#L103), `intent` can be `None` if it's missed and that gets translated to an empty list of `mapping_intents` which later means `intent_not_blocked` is `True` as long as the `not_intent` doesn't contain the latest user message intent. In a nutshell, it means that if the `intent` parameter is not used, then the slot can be filled with the `value` for any given intent as long as it's not listed as `not_intent`. I can rewrite this as you suggested.

```yaml-rasa
slots:
  slot_name:
    type: any
    mappings:
    - type: from_intent
      value: my_value
      intent: intent_name
      not_intent: excluded_intent
```

#### from_trigger_intent

The `from_trigger_intent` mapping will fill slot `slot_name` with value `my_value`
if a form is activated by a user message with intent `intent_name`.
The slot mapping will not apply if the intent of the message is
`excluded_intent`. Note that you can
also define lists of intents for the parameters `intent` and `not_intent`.

A slot with `from_trigger_intent` mapping must always include mapping conditions

> and be listed in the form's `required_slots`.

Author: I'm actually going to change my mind about the first half of the sentence after rechecking the code, doesn't need to include mapping conditions, only needs to be in the form's required slots. This can also be checked with `rasa data validate`, should I link to that info instead?

> and be listed in the form's `required_slots`.

Reviewer 2: Yes, that makes sense!

slots:
  slot_name:
    type: any
    mappings:
    - type: from_trigger_intent
      value: my_value
      intent: intent_name
      not_intent: excluded_intent
      conditions:
      - active_loop: your_form
forms:
  your_form:
    required_slots:
    - slot_name
```

### Mapping Conditions

If you prefer to extract and set a slot only within the context of a form, you can indicate
the name of the form in the `conditions` key of a mapping. Conditions list the form name(s)
for which the mapping is applicable in the `active_loop` key.

Conditions can also include the name of the `requested_slot`. If `requested_slot` is not mentioned,
then the slot will be set if relevant information is extracted, regardless of which slot is being
requested by the form.

```yaml-rasa
slots:
  slot_name:
    type: text
    mappings:
    - type: from_text
      intent: intent_name
      conditions:
      - active_loop: your_form
        requested_slot: slot_name
      - active_loop: another_form
```

:::note
If conditions are not included in a slot mapping, the slot is eligible to be set universally
outside the context of any forms. However if a slot is also listed in a form's `required_slots`,

> the slot can also be requested by the form if unset at that moment.

Author: I wanted to emphasise the behaviour without `conditions` so that there are no surprises if this key is omitted. And also to point out that forms cannot update slots, they can only set slots if they are unset at that point in the conversation.

:::

### Custom Slot Mappings

A custom slot mapping is useful when none of the predefined mappings fit your use case.
For example, if you use custom policies, custom actions or the custom validation action to extract and validate slots,

> you will have to define the slot mapping in the domain as `custom`.

Author: >  not sure how a policy could be setting slots using this method?

I don't know either, but I heard feedback from some Rasa superheroes that some custom components they use can set slots too.

> it reads as: if the user already uses custom slot mappings, then they need to specify custom slot mappings.

Mhm, yes, I think I was subconsciously writing it as a migration guide for an audience of previous Rasa developers vs. new developers. I'll rephrase as suggested.

In order to write your own extraction and / or validation code, we recommend you to create a custom action inheriting
from the `ValidationAction` interface from the Rasa SDK and defining the `extract_<slot_name>` and / or
`validate_<slot_name>` methods. The name of the action is already implemented and
should not be overridden: `action_validate_slot_mappings`. Please make sure to add this action name to the list of

> actions in the domain.

Author: It's only an option, the alternative is the custom action example. I'll attempt to rephrase this more clearly.

## Config

The `config` key in the domain file maintains the `store_entities_as_slots` parameter.
When an entity is recognized by the NLU model and it matches the `from_entity` slot mapping, `store_entities_as_slots`
defines whether the entity value should be placed in that slot.
By default, this behaviour is switched on.

> You can turn off this functionality by setting the `store_entities_as_slots` parameter to `false`:

Reviewer 2: with `auto_fill` the turning off made sense, but in the new scenario, what's the point of controlling this here? Doesn't it implicitly nullify the `from_entity` mappings, so one could just as well delete them?
I'm sure there's a good reason, just think we should tell users what it is

Author: Yes, you're right, this doesn't make sense that much anymore; i only kept it because this attribute (`store_entities_as_slots`) conditions the running of a Domain method (`slots_for_entities()`) that gets called in `StoryStep.explicit_events`. So it helps with a story such as the one debated in a thread a while ago:
```
- story: example story
  steps:
  - or:
     - intent: greet
     - intent: greet
        entities:
        - name: Molly
 - action: utter_greet
```
Probably an idea to mention that if this parameter gets turned off the stories won't work accordingly 🤔

Reviewer 2: Ah right, I remember that thread now. 

Not sure this is clarified yet though: Does `store_entities_as_slots: false` == removing all `from_entity` mappings?
Or, is there some way the `from_entity` mappings get applied, even when `store_entities_as_slots: false`?

 Does this parameter control how the _syntax_ in stories is interpreted at training time, and is this any different from how user messages are interpreted & used to set slots at inference time?
e.g.
if `store_entities_as_slots: true`, then this story and the following one are equivalent:
```
- story: example story
  steps:
  - or:
     - intent: greet
     - intent: greet
        entities:
        - name: Molly
 - action: utter_greet
```
==
```
- story: example story
  steps:
  - or:
     - intent: greet
     - intent: greet
        entities:
        - name: Molly
     - slot_was_set:
        - name: Molly
 - action: utter_greet
```

If `store_entities_as_slots: false`, then the first story is rather equivalent to this other one:

```
- story: example story
  steps:
  - or:
     - intent: greet
     - intent: greet
        entities:
        - name: Molly
     - slot_was_set:
        - name: null
 - action: utter_greet
```

Reviewer 2: An example of the impact of this parameter is definitely needed here, I've ran into A Lot of user confusion on this behaviour.

Author: > Does store_entities_as_slots: false == removing all from_entity mappings?
> Or, is there some way the from_entity mappings get applied, even when store_entities_as_slots: false?

No, this will not affect how `from_entity` slots are filled during the conversation (meaning the default action will carry on running just fine).

> Does this parameter control how the syntax in stories is interpreted at training time, and is this any different from how user messages are interpreted & used to set slots at inference time?

Yes, this parameter will only control to how stories are interpreted at training time. It doesn't influence how slots are set at inference time after every user turn.

I don't know for sure if that exact story syntax you listed for `store_entities_as_slots: false` works, but  it's in the right direction: if this parameter gets turned off, then the tracker won't have a `SlotSet` event for the entity in the user message listed in the story.

Reviewer 2: Ok! Then that's what I think we should explain in this section,  that this parameter is only relevant to reading stories and turning them into trackers, and making slot set events out or entities.
